lib/SILOptimizer/Transforms/SILMem2Reg.cpp

```diff
@@ -577,45 +583,58 @@ StoreInst *StackAllocationPromoter::promoteAllocationInBlock(
+                                           runningVals->value);
+            } else {
+              // If we dont't have enough information, end it
+              // endLexicalLifetime.
+              assert(!deinitializationPoints[blockPromotingWithin]);
+              deinitializationPoints[blockPromotingWithin] = li;
+            }
           }
+          if (runningVals)
+            runningVals->isStorageValid = false;
+          if (lastStoreInst)
+            lastStoreInst->isStorageValid = false;
         }
-        if (runningVals)
-          runningVals->isStorageValid = false;
-        if (lastStoreInst)
-          lastStoreInst->isStorageValid = false;
-      }
-      if (runningVals) {
-        // If we are loading from the AllocStackInst and we already know the
-        // content of the Alloca then use it.
-        LLVM_DEBUG(llvm::dbgs() << "*** Promoting load: " << *li);
-        replaceLoad(li, runningVals->value.replacement(asi), asi, ctx, deleter,
-                    instructionsToDelete);
-        ++NumInstRemoved;
-      } else if (li->getOperand() == asi &&
-                 li->getOwnershipQualifier() != LoadOwnershipQualifier::Copy) {
-        // If we don't know the content of the AllocStack then the loaded
-        // value *is* the new value;
-        // Don't use result of load [copy] as a RunningVal, it necessitates
-        // additional logic for cleanup of consuming instructions of the result.
-        // StackAllocationPromoter::fixBranchesAndUses will later handle it.
-        LLVM_DEBUG(llvm::dbgs() << "*** First load: " << *li);
-        runningVals = {LiveValues::toReplace(asi, /*replacement=*/li),
-                       /*isStorageValid=*/true};
+
+        if (runningVals) {
+          // If we are loading from the AllocStackInst and we already know the
+          // content of the Alloca then use it.
+          LLVM_DEBUG(llvm::dbgs() << "*** Promoting load: " << *li);
+          replaceLoad(inst, runningVals->value.replacement(asi), asi, ctx,
+                      deleter, instructionsToDelete);
+          ++NumInstRemoved;
+        } else if (li->getOperand() == asi &&
+                   li->getOwnershipQualifier() !=
+                       LoadOwnershipQualifier::Copy) {
+          // If we don't know the content of the AllocStack then the loaded
+          // value *is* the new value;
+          // Don't use result of load [copy] as a RunningVal, it necessitates
+          // additional logic for cleanup of consuming instructions of the
+          // result. StackAllocationPromoter::fixBranchesAndUses will later
+          // handle it.
+          LLVM_DEBUG(llvm::dbgs() << "*** First load: " << *li);
+          runningVals = {LiveValues::toReplace(asi, /*replacement=*/li),
+                         /*isStorageValid=*/true};
+        }
+        continue;
+      } else {
+        auto *lbi = cast<LoadBorrowInst>(inst);
+        if (runningVals) {
```

Comment:
I'm just trying to understand the situation that the original code handles. If there's no running value, then we leave the original `load_borrow` in place? Does that mean that we may partially promote an `alloc_stack` to a value? Do we leave in stores in that case?

Reply:
Yes, if there is no running value, we will leave the load_borrow as is. This function is called from `StackAllocationPromoter::pruneAllocStackUsage` which does not fully remove the uses of the alloc_stack. Remaning uses will be replaced in `StackAllocationPromoter::promoteAllocationToPhi`